Introduce WebhookEventType

sphinks opened the pull request:
- webhook event type in code of webhook class
- tests for webhook class

sphinks committed:
- introduce WebhookEventType
- modify tests
- correct readme according to new way of using webhook event type class

sphinks committed:
- fix tests names

sphinks committed:
- use event type Unknown
- add event type for tests
- correct tests

sphinks committed:
- fix misprints in comments

sphinks commented:
@bob983 any update on this?:)

bob983 commented:
Hey @sphinks, I thought you were still working on it, my bad!

sphinks commented:
@bob983 if you see anything to add in scope of this PR - let me know, please.

sphinks committed:
- refactor http methods in separate class

sphinks commented:
Add some refactor in the same PR.
- move http wrapper functions to separate class

sphinks committed:
- revert move of Jsonparse method to ApiClient

bob983 commented on src/main/java/net/helpscout/api/ApiClient.java:
```diff
@@ -1442,52 +1429,23 @@ private void setThreadProperties(ConversationThread thread) {
-            throw new RuntimeException(ex);
-        }
-    }
+    private <T> ArrayList<T> getPageItems(String url, Class<T> clazzType) throws ApiException {
```
I think it would be better if the return type was `List<T>` and not `ArrayList<T>`, also `clazzType` could be named after what it really represents, for example `responseType`

sphinks replied:
Just move this function from one position to another without refactoring it. Thanks for good notice, I see some similar functions, will fix them all this way.

bob983 commented on src/main/java/net/helpscout/api/ApiClient.java:
```diff
@@ -1442,52 +1429,23 @@ private void setThreadProperties(ConversationThread thread) {
+    private JsonElement parseJson(String url, String json) {
+        LoggerFactory.getLogger(getClass()).trace("{}: {}", url, json);
+        JsonElement obj = (new JsonParser()).parse(json);
+        return obj;
```
I think you can return result raight away:

```
return new JsonParser().parse(json);
```

Also, does the `JsonParser` need to be created every time?

sphinks replied:
Also, just moved this function, so do not pay attention at this moment. But good point about creating Parser every time - will check it. To tell the truth I consider this class is not right place for this function (Util function), but it is still here cause all functions that use it, are still tightly coupled with ApiClient class (and parseJson performing logging here). So I suppose, once I have a chance to decouple this functions I will move it in JsonUtils.

bob983 commented on src/main/java/net/helpscout/api/HTTPMethodWrapper.java:
```diff
@@ -0,0 +1,70 @@
+    private String apiKey;
+    private String baseUrl;
+
+    public HTTPMethodWrapper() {
```
This is not necessary, if there's no other constructor, compiler will add the default one

sphinks replied:
You are right, I was starting with non-default constructor, but then comes to this variant with setters and forget to clean constructor.

bob983 commented on src/main/java/net/helpscout/api/HTTPMethodWrapper.java:
```diff
@@ -0,0 +1,70 @@
+    public HTTPMethodWrapper() {
+    }
+
+    public void setApiKey(String apiKey) {
```
You could use `@Setter` annotation on the class level and you'll get your setters generated by lombok

sphinks replied:
Cool hint! Thanks, not yet know all tricks of lombok.:)

bob983 commented on src/main/java/net/helpscout/api/HTTPMethodWrapper.java:
```diff
@@ -0,0 +1,70 @@
+
+    public <T> T doPost(String url, String requestBody, int expectedCode, ResultExtractor<T> extractor) throws ApiException {
+
+        try (HTTPConnectionWrapper conn = new HTTPConnectionWrapper(apiKey, baseUrl + url, METHOD_POST, expectedCode, requestBody)) {
```
All the methods follow the same `try/catch/catch` format, is there a way to extract this functionality?

sphinks replied:
Hm, maybe...

sphinks replied:
Take a closer look, I do not see easy way to eliminate duplication of try/catch, due to different processing of result and returned type. If have any idea how to improve this part of code - let me know :)

bob983 commented on src/main/java/net/helpscout/api/cbo/WebhookEventType.java:
```diff
@@ -0,0 +1,86 @@
+    private final int value;
+    private final String label;
+
+    private WebhookEventType(int value, String label) {
```
You could use `@AllArgsConstructor` and `@Getter` and remove the constructor and getters.

sphinks replied:
One more lombok hint:) Thanks!

bob983 commented:
@sphinks I took me a while to get to the PR, sorry. I added some comments, all of them are about low level choices, nothing serious. Please have a look. Also, did you get your t-shirt? :)

sphinks commented:
@bob983 thanks for some great hints about lombok!:) Will try to improve the code a little bit. T-Shirt does not arrive, but I suppose it should be soon at my post office. :-)

sphinks committed:
- delete unused parameter in ApiClient.getPage
- JsonParser create once per ApiClient object
- WebHookEventType use lombok annotations

sphinks committed:
- use lombok annotations for getters in HttpMethod Wrapper

sphinks committed:
- use Setter annotations of lombok in HttpMethodWrapper

sphinks committed:
- getters not needed in HtppMethodWrapper

sphinks commented:
@bob983 have some time today and improved PR a little bit. Take a look, there is one place I leave as is, if you also do not see easy way to improve it, then we can finish with this PR.

bob983 commented:
Looking good, @sphinks 
I'll run the tests, have a second look and hopefully merge it soon.

sphinks commented:
@bob983 great! :+1:

sphinks commented:
Hi @bob983,
any update on merging? I'm just prepared new PR and waiting this to be merged. :)

sphinks commented:
@bob983 Are still here?:)

bob983 commented:
Oh my, the time flies, sorry :( ... Let me dig into this one now.

bob983 commented on src/main/java/net/helpscout/api/ApiClient.java:
```diff
@@ -37,34 +37,32 @@
+    private final JsonParser jsonParser = new JsonParser();
+
+    private ApiClient() {
+        httpMethodWrapper.setBaseUrl(DEFAULT_BASE_URL);
```
Could you please move this into the `HTTPMethodWrapper` constructor?

So rather than 

``` java
private static final HTTPMethodWrapper httpMethodWrapper = new HTTPMethodWrapper();
private ApiClient() {
      httpMethodWrapper.setBaseUrl(DEFAULT_BASE_URL);
}
```

this might be better:

``` java
private static final HTTPMethodWrapper httpMethodWrapper = new HTTPMethodWrapper(DEFAULT_BASE_URL);
private ApiClient() {
// constructor is empty, could be deleted actually    
}
```

sphinks replied:
Hi @bob983 
Sorry that did not answer for a long time. I was on vacation, so I just returned. 

Actually it is easy change, it can be done for sure. But I'm confused with the idea that service class, will take a care of some specific parameter of system. I mean that URL of api is really belongs to ApiClient class by logic and it is as for me bad move to spread system specific parameter to service class, which should not know anything about system.

One more point - I would not also delete a private contructor of ApiClient, cause we are using here Singelton. So actually it is not a change that can be done for sure from my point of view.

What do you think about it?

bob983 replied:
Well, you're passing the `DEFAULT_BASE_URL` to `HTTPMethodWrapper` anyway, it's just about the way it's passed there. Right now it's a setter and I don't like that. The `HTTPMethodWrapper` is created unconfigured (== it can't work properly) and then later it's configured by constructor.

And yep, you are right about the constructor, but it can be at least replaced by
`@NoArgsConstructor(access= AccessLevel .PRIVATE)`

bob983 reviewed (changes requested):
All good, just one note about the `ApiClient` constructor.

sphinks committed:
- Create HTTPMethodWrapper with Base_url parameter

sphinks commented:
@bob983 have implemented what you asked me for, sorry for delay.
After your second commend I understand that I get you idea wrong, nevertheless you give a code snippet (reading that time from mobile phone). :) Initially I consider that you ask me to me `DEFAULT_BASE_URL` constant to `HttpMethodWrapper` but leave constructor as it was. That is why I was talking about letting service method to know additional information which does not related to this class. Now I get that I was wrong and you just supposed to inject this parameter directly into constructor. :) That is OK. I should agree from very begining. :)

sphinks commented:
@bob983 let's merge it at last ;-)

bob983 reviewed (approved):
All good, thanks!

sphinks commented:
@bob983 Yeah! Rock this merge! :)
